### PR TITLE
Added finding quantiles on GPU.

### PR DESCRIPTION
**Pull request by canonizer**
- this includes datasets where weights are assigned to data rows
- as the quantiles found by the new algorithm are not the same
  as those found by the old one, test thresholds in
    tests/python-gpu/test_gpu_updaters.py have been adjusted.

### CODE REVIEW

File: src/tree/param.h [hunk 1]
```diff
@@ -76,6 +76,8 @@ struct TrainParam : public dmlc::Parameter<TrainParam> {
   int gpu_id;
   // number of GPUs to use
   int n_gpus;
+  // number of rows in a single GPU batch
```
Comment (RAMitchell) [inline src/tree/param.h]: This parameter seems to have crept back in since your last PR.

**Comment by hcho3** [comment]
> I also think the existing quantile code (WXQSketch, gmat, hmat) is probably long overdue a refactor. It is difficult to understand and work with. Perhaps @hcho3 can help me with this in a future PR.

I agree. The quantile generation is important for distributed training as well as for the 'hist' algorithm. Let me review literature on weighted quantile generation. Also, the quantized matrix generation needs clean-up as well. I've been rather occupied lately in getting started with MXNet contribution, but I'll try to make more time for XGBoost.

Ps. I'll be mentoring a new intern this summer to help him contribute to XGBoost. He is planning to improve distributed training.

File: src/common/device_helpers.cuh [hunk 2]
```diff
@@ -252,6 +271,18 @@ T1 DivRoundUp(const T1 a, const T2 b) {
   return static_cast<T1>(ceil(static_cast<double>(a) / b));
 }
```
[the remark this answers is not in the document]
Comment (canonizer) [reply src/common/device_helpers.cuh]: Done.

File: src/tree/updater_gpu_hist.cu [hunk 3]
```diff
@@ -348,41 +330,50 @@ struct DeviceShard {
...
-    thrust::device_vector<float> cuts_d(hmat.cut);
-    thrust::device_vector<size_t> cut_row_ptrs_d(hmat.row_ptr);
-
-    // find the maximum row size
-    thrust::device_vector<size_t> row_ptr_d(
-        row_batch.offset.data() + row_begin_idx, row_batch.offset.data() + row_end_idx  + 1);
-
-    auto row_iter = row_ptr_d.begin();
+    row_ptrs.resize(n_rows + 1);
+    thrust::copy(&row_batch.offset[row_begin_idx], &row_batch.offset[row_end_idx + 1],
```
Comment (RAMitchell) [inline src/tree/updater_gpu_hist.cu]: `&row_batch.offset[row_end_idx + 1]`

This is actually an array out of bounds error if the compiler does not optimise. MSVC will crash in debug mode on the statement. Instead use:
`row_batch.offset.data() + row_end_idx  + 1`

Also regarding thrust copy, I ran into issues with this creating large temporary memory buffers when copying nontrivial types (GradientPair). It should be fine for primitive types but I was considering banning thrust::copy completely via some kind of lint rule. Any thoughts on this?

**Comment by canonizer** [comment]
Fixed the comment and the Python test failure.

**Pull request merged**